Upgrade react, react-bootrstrap and react-router (resolve #429)

epaminond opened the pull request:
This resolves https://github.com/botpress/v12/issues/210: `react`, `react-router`, `react-bootstrap` and other related packages were updated here. Please also check relevant PR-s to other modules:

- [x] https://github.com/botpress/botpress-platform-webchat/pull/42 ❗
- [x] https://github.com/botpress/enterprise/pull/14 ❗
- [ ] https://github.com/botpress/botpress-messenger/pull/79
- [ ] https://github.com/botpress/botpress-telegram/pull/4
- [ ] https://github.com/botpress/botpress-api.ai/pull/24
- [ ] https://github.com/botpress/botpress-nlu/pull/4❗
- [ ] https://github.com/botpress/botpress-slack/pull/24
- [x] https://github.com/botpress/botpress-skill-choice/pull/1 ❗
-  ❌ https://github.com/botpress/botpress-wit/pull/8
-  ❌ https://github.com/botpress/botpress-rivescript/pull/9
- [ ] https://github.com/botpress/botpress-twilio/pull/6
- ❌https://github.com/botpress/botpress-sms/pull/1 (delete this repo)
- ❌ https://github.com/botpress/botpress-kik/pull/1 (delete this repo)
-  ❌ https://github.com/botpress/botpress-rasa/pull/2
- [ ] https://github.com/botpress/botpress-audience/pull/3  ❗
- ❌ https://github.com/botpress/botpress-webhook/pull/2
- [ ] https://github.com/botpress/botpress-terminal/pull/1  ❗
- [ ] https://github.com/botpress/botpress-scheduler/pull/6  ❗
- ❌https://github.com/botpress/botpress-subscription/pull/4 (deprecated since User Tags, won't merge)
- [ ] https://github.com/botpress/botpress-hitl/pull/10 ❗
- [ ] https://github.com/botpress/botpress-broadcast/pull/6 ❗
- [ ] https://github.com/botpress/botpress-analytics/pull/11 ❗

I'm still planning to retest things again on Monday, but if you notice issues yourself, let me know!

emirotin commented on src/web/components/Content/Select/index.js:
```diff
@@ -155,11 +157,12 @@ class SelectContent extends Component {
+              key={i}
               onClick={() => this.setCurrentCategory(category.id)}
-              className="list-group-item list-group-item-action"
+              className={`list-group-item list-group-item-action ${i === this.state.activeItemIndex ? 'active' : ''}`}
```
could use `classnames` here

epaminond replied:
Fixed that!

emirotin commented on src/web/components/Layout/Sidebar.jsx:
```diff
@@ -87,59 +81,35 @@ class Sidebar extends React.Component {
-      <li key={`menu_module_${module.name}`} className={classNames}>
-        <Link to={path} title={module.menuText}>
+      <li key={`menu_module_${module.name}`}>
+        <NavLink to={path} title={module.menuText} activeClassName={`${style.active}`}>
```
string interpolation unneeded

emirotin replied:
same below

epaminond replied:
Fixed that!

emirotin commented on src/web/views/FlowBuilder/common/action.jsx:
```diff
@@ -13,23 +13,22 @@ class ActionItem extends Component {
-    this.fetchItem(this.props.text)
+    const itemId = this.textToItemId(this.props.text)
+    this.state = { itemId }
+    this.fetchItem(itemId)
```
It's usually recommended to have side-effects in componentDidMount

epaminond replied:
Fixed that!

emirotin commented on src/web/views/FlowBuilder/common/action.jsx:
```diff
@@ -13,23 +13,22 @@ class ActionItem extends Component {
+  fetchItem = itemId => {
     if (itemId) {
-      this.props.fetchContentItem(itemId).then(this.props.refreshFlowsLinks)
+      this.props.fetchContentItem(itemId).then(() => this.props.refreshFlowsLinks())
```
Why is this change needed? `refreshFlowsLinks` is unbound anyway

epaminond replied:
Not sure. Possibly it was left after some debugging.

emirotin commented on src/web/views/FlowBuilder/sidePanel/properties/ActionSection.jsx:
```diff
@@ -105,10 +105,10 @@ export default class ActionSection extends Component {
     return (
       <div>
-        <Panel collapsible defaultExpanded={true} header={header}>
+        <Panel collapsible="true" defaultExpanded={true} header={header}>
```
should it be `{true}` instead?

emirotin replied:
same below

epaminond replied:
@emirotin , it it's `{true}`, I'm getting warning:

```
Warning: Received `true` for a non-boolean attribute `collapsible`.
If you want to write it to the DOM, pass a string instead: collapsible="true" or collapsible={value.toString()}.
```

emirotin replied:
Oh, I thought it's the `Panel`'s prop, makes sense otherwise

epaminond committed:
Upgrade react, react-bootrstrap and react-router (resolve #429)

epaminond commented:
@slvnperron , I've upgraded most other modules and added links to relevant pull-requests to the top message. I've also found a few minor changes in `react-bootstrap` `Panel` api and fixed them across all the packages and `botpress` itself. Let me know if you see any other issues that I could miss.

slvnperron commented:
@epaminond Thanks, that looks great! Thanks for sifting and updating all the repos.

Like discussed over the phone yesterday, the new release will require a couple of repo naming changes. How do you suggest we go about merging these pull requests and make them only compatible with the upcoming Botpress release?

I'm wondering if we should adopt the mono-repo approach for all the official modules, as this would greatly reduce the maintenance burden?  @epaminond @emirotin

epaminond commented:
@slvnperron , I'll check how we can utilize https://github.com/botpress/botpress-version-manager for that.

As for monorepo - I think this is a matter of preference. It has both pros and cons but indeed updating multiple modules in single repo would be easier this time.

epaminond commented:
@slvnperron , we can set compatible `botpress` version in `version-manager`, but we could also utilize `peerDependencies` for the same thing. With `peerDependencies` we could also set compatible versions of other packages like `botpress-messenger` for `botpress-hitl` and this is a native solution.

So I can suggest the following upgrade path:
1. We upgrade `botpress` core module and publish it to npm with major version change
2. We keep modules with old names in npm either making them incompatible with newer `botpress` version via `version-manager` and/or `peerDependencies` and adding a note to the README
3. We upgrade modules and publish them with new names (if needed) making them compatible with newer `botpress` version via `version-manager` and/or `peerDependencies`

slvnperron commented:
Thanks Alex, your path sounds good. Let's also publish the new modules
under new names:

Channels with: botpress-channel-* (instead of platform)
Skills with: botpress-skills-* (notice the addition of `s` to skill, which
we have to also change in botpress, i.e. we need to make modules able of
exporting multiple skills)



On Thu, Feb 15, 2018 at 9:58 AM, Alex Pletnov <notifications@github.com>
wrote:

> @slvnperron <https://github.com/slvnperron> , we can set compatible
> botpress version in version-manager, but we could also utilize
> peerDependencies for the same thing. With peerDependencies we could also
> set compatible versions of other packages like botpress-messenger for
> botpress-hitl and this is a native solution.
>
> So I can suggest the following upgrade path:
>
>    1. We upgrade botpress core module and publish it to npm with major
>    version change
>    2. We keep modules with old names in npm either making them
>    incompatible with newer botpress version via version-manager and/or
>    peerDependencies and adding a note to the README
>    3. We upgrade modules and publish them with new names (if needed)
>    making them compatible with newer botpress version via version-manager
>    and/or peerDependencies
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/botpress/botpress/pull/436#issuecomment-365951968>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQStB1MU8JmEunWO1wwCkRD7-6PyMpYks5tVEX7gaJpZM4SBGDM>
> .
>



-- 
*Sylvain Perron*
Director at Botpress <https://botpress.io/> | ⚡ the conversational
revolution
sylvain@botpress.io
T | +1 418 440 7586
Calendar Slots: Partnerships
<https://calendar.google.com/calendar/selfsched?sstoken=UUNxa1dWY0VQUTdDfGRlZmF1bHR8YTFkODZkYmE5ZTE0MmVkOGZkMjFkNjU3MGEwYjgxZjc>
 | Enterprises
<https://calendar.google.com/calendar/selfsched?sstoken=UUlrdlNiM3dGblEtfGRlZmF1bHR8Mzk5MDUwMGQzZWM1NDMwMjg2M2UyY2QzM2EwNWE1ODg>

emirotin commented:
I've used monorepo approach several times, it's easier indeed when you have many similar modules. You can make the identical changes in a single PR. I've been using `lerna` (extracted from babel as far as I remember) and it worked pretty well. Also there's a related thing called `yarn workspaces` but I've never used it myself. It's said to work well with lerna though.

epaminond commented:
@slvnperron , do I have access rights to publish packages at NPM?

slvnperron commented:
@emirotin Wow thanks, lerna is f** awesome!

@epaminond Not yet, will do now :)

slvnperron commented:
@epaminond What's your npm username?

epaminond commented:
@slvnperron , it's "epaminond"

slvnperron commented:
@epaminond Can we solve the merge conflicts and push this to a x/react16 branch? Would like to test a bot with the following:

- NLU
- Enterprise
- HITL
- Broadcast
- Scheduler
- Audience
- Analytics
- Terminal
- Skill Choice

(These repos should come from the @botpress/ modules)

Once that's tested properly let's merge in `x`

epaminond commented:
@slvnperron , done!

epaminond commented:
@slvnperron , should I also include updates introduced in https://github.com/botpress/botpress/pull/460 to `x/react16` branch?

epaminond committed:
Fix adding flow link points
